### PR TITLE
[POC][TF FE] Support SavedModel format (with compression)

### PR DESCRIPTION
**Pull request by gkrivor**
### Details:
 - Added support of TF Saved Model file format (with compression support)

### Tickets:
 - 101921, ...

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
Added label `don't_merge` to understand aspects of new dependency

File: src/core/src/op/util/scatter_nd_base.cpp [hunk 1]
```diff
@@ -35,7 +35,7 @@ void ov::op::util::ScatterNDBase::validate_and_infer_types() {
     element::Type updates_et = get_input_element_type(UPDATES);
 
     NODE_VALIDATION_CHECK(this,
-                          indices_et == element::i32 || indices_et == element::i64,
+                          indices_et == element::i32 || indices_et == element::i64 || indices_et == element::dynamic,
```
Comment (slyalin) [inline src/core/src/op/util/scatter_nd_base.cpp:38]: Revert this. It is already fixed in master.

**Review by ilya-lavrenov — APPROVED** [review]

File: src/frontends/tensorflow_common/include/helper_ops/string_constant.hpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#include <vector>
+
+#include "internal_operation.hpp"
+
+namespace ov {
+namespace frontend {
+namespace tensorflow {
+
+/// Pseudo-entity for storing strings
+class StringConstant : public InternalOperation {
```
Comment (rkazants) [inline src/frontends/tensorflow_common/include/helper_ops/string_constant.hpp:17]: Let us inherit this class from `UnsupportedConstant`. So there will be no need to update `ConstToResultRemover`

File: src/frontends/tensorflow_common/src/op/identity.cpp [hunk 3]
```diff
@@ -14,7 +14,13 @@ namespace tensorflow {
 namespace op {
 
 OutputVector translate_identity_op(const NodeContext& node) {
-    vector<string> supported_ops = {"Identity", "PreventGradient", "Snapshot", "StopGradient"};
+    vector<string> supported_ops = {"Identity",
+                                    "PreventGradient",
+                                    "Snapshot",
+                                    "StopGradient",
+                                    "ReadVariableOp",
+                                    "ShardedFilename",
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/identity.cpp:22]: I think for `ShardedFilename` we need to implement separate translator. We expect `StringConstant` input there and this translator will replace `ShardedFilename` with the new `StringConstant` that is computed as `%s-%05d-of-%05d, basename, shard, num_shards` according to its specification. Let make it in the straightforward way.

**Review by rkazants — COMMENTED** [review]
@andrei-kochin, @gkrivor, also we definitely need tests

**Comment by rkazants** [comment]
> > @andrei-kochin, @gkrivor, also we definitely need tests
> 
> agree.
> 
> But not to block the PR, we can add tests after FF.
> 
> We need to merge this PR rather earlier than later to handle distribution aspects because of new dependency.

Please create a ticket before merge. And let us plan this for this release as must have.
Now it is treated as POC.

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+            std::map<std::string, const ::tensorflow::SignatureDef*> validSignatures = {};
+            for (const auto& sit : meta_graph.signature_def()) {
+                const std::string& key = sit.first;
+                const ::tensorflow::SignatureDef& val = sit.second;
+                if (is_valid_signature(val)) {
+                    validSignatures[key] = &val;
+                }
+            }
+
```
[the remark this answers is not in the document]
Comment (gkrivor) [reply src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:234]: Not sure it should be commented. It is different entities. Signatures != tags.

I have a model which has a __saved_model_init_op signature. I didn't find a meaningful usage for that.

Also I can refer to comment:
https://github.com/tensorflow/tensorflow/blob/dad0d43036e56b0ab538cc0a46514742c6c74857/tensorflow/compiler/mlir/tensorflow/translate/import_model.cc#L4076

File format doesn't have a limitations for signature amount, they might be added later.

File: src/frontends/tensorflow/src/decoder_proto.cpp [hunk 5]
```diff
@@ -194,9 +203,22 @@ ov::Any DecoderProto::get_attribute(const std::string& name) const {
...
+                ov_type.is_static(),
+                "Encountered unknown element type " + DataType_Name(tf_type) + " on an empty tensor_proto");
+        } else {
+            ov_type = ov::element::u64;
+            pshape.resize(0);
+            pshape.push_back(tensor_proto.string_val_size());
+        }
+        if (tf_type == ::tensorflow::DataType::DT_STRING) {
+            auto data = std::vector<std::string>();
+            for (auto& item : tensor_proto.string_val()) {
```
[the remark this answers is not in the document]
Comment (gkrivor) [reply src/frontends/tensorflow/src/decoder_proto.cpp:217]: ok

File: src/frontends/tensorflow/src/decoder_proto.cpp [hunk 6]
```diff
@@ -168,7 +173,11 @@ ov::Any DecoderProto::get_attribute(const std::string& name) const {
         if (list.type_size()) {
             std::vector<ov::element::Type> res;
             for (int idx = 0; idx < list.type_size(); ++idx) {
-                res.emplace_back(get_ov_type(list.type(idx)));
+                if (list.type(idx) != ::tensorflow::DataType::DT_STRING) {
+                    res.emplace_back(get_ov_type(list.type(idx)));
+                } else {
+                    res.emplace_back(ov::element::undefined);
```
[the remark this answers is not in the document]
Comment (gkrivor) [reply src/frontends/tensorflow/src/decoder_proto.cpp:179]: It is place for some DT_RESOURCE or other unsupported types... But if you think it is better - np.

File: src/frontends/tensorflow/src/decoder_proto.cpp [hunk 7]
```diff
@@ -194,9 +203,22 @@ ov::Any DecoderProto::get_attribute(const std::string& name) const {
...
         auto ov_type = get_ov_type(tf_type);
-        FRONT_END_GENERAL_CHECK(
-            ov_type.is_static(),
-            "Encountered unknown element type " + DataType_Name(tf_type) + " on an empty tensor_proto");
+        if (tf_type != ::tensorflow::DataType::DT_STRING) {
+            FRONT_END_GENERAL_CHECK(
+                ov_type.is_static(),
+                "Encountered unknown element type " + DataType_Name(tf_type) + " on an empty tensor_proto");
+        } else {
+            ov_type = ov::element::u64;
```
[the remark this answers is not in the document]
Comment (gkrivor) [reply src/frontends/tensorflow/src/decoder_proto.cpp:211]: Good catch! It is code from old WA, I've removed it.

File: src/frontends/tensorflow_common/src/op/identity.cpp [hunk 8]
```diff
@@ -14,7 +14,13 @@ namespace tensorflow {
...
+                                    "MergeV2Checkpoints"};
```
[the remark this answers is not in the document]
Comment (gkrivor) [reply src/frontends/tensorflow_common/src/op/identity.cpp:23]: I'll try

File: src/frontends/tensorflow_common/src/op/identity.cpp (hunk 3 above)
Comment (gkrivor) [reply src/frontends/tensorflow_common/src/op/identity.cpp:22]: Right now it is unused. As I understand - leave it as is right now.

**Review by rkazants — APPROVED** [review]

**Pull request merged**